platform/api.templates/src/org/netbeans/api/templates/CreateFromTemplateHandler.java

```diff
@@ -59,4 +64,171 @@
             CreateDescriptor    desc
     ) throws IOException;
 
+    /**
+     * Replaces ${param} tokens in a String. The parameter reference has syntax "${" paramName [":" defaultValue ] "}".
+     * Parameters are replaced recursively. "${" can be escaped by \, so "\${" will not act as parameter reference start.
+     * @param expression string template
+     * @param parameters parameter values
+     * @return evaluated string.
+     * @since 1.23
+     */
+    public static String mapParameters(String expression, Map<String, ?> parameters) {
```

Comment:
There is `org.openide.util.MapFormat` - if the suggestion to use `freemarker` isn't going to fly - wouldn't it be better to reuse the existing `MapFormat` or expand it?

Reply:
Reuse: I don't think its's possible 'as is' as I'd like to keep the default - which gives some troubles if the default itself contains delimiters (see this impl). Also the replacement in MapFormat is not recursive - that could be also solved by the caller by calling MapFormat repeatedly, but could be added.

If you think it's worth moving to MapFormat, I'll try; but I'd like to retain the `${...}` syntax, which is consistent with Ant, Freemarker or other interpolation engines.

Reply:
I still like the [use of the templating engine itself](https://github.com/apache/netbeans/pull/3262#discussion_r740114513) more than our own substitution mechanism...